### PR TITLE
Remove lifetime constraints from `wgpu::ComputePass` methods

### PR DESCRIPTION
**Pull request by Wumpf**
**Connections**
* Follow-up & based on #5432
     * ⚠️ Will rebase once #5432 lands as this contains all the changes again.
* Important puzzle piece of #1453

Part of a series towards lifetime removal on compute passes:
* #5409
* #5432
* ➡️ #5570
* #5575
* #5620
* #5671

**Description**
**Removes the need for lifetime dependency on any resource passed into a ComputePass.**
Lifetime constraints to the originating encoder & queries used during creation still apply for the moment (to be addressed in a follow-up.

Directly builds on #5432 by moving the creation of `ArcComputeCommand` into compute pass methods.

To accomplish this I had to add a HalApi generic dependency on `ComputePass`. Since I didn't want to introduce a new object to the id/objecttracking system, I instead decided to create a `DynComputePass` on `wgpu/wgpu_core.rs` side. Hopefully we can simplify this as part of the generic refactor (#5124). Note that the new lifetime dependency made boxing of the wgpu-core compute pass is unavoidable at this point and `DynComputePass` allows us to merge the otherwise required `gfx_select` & type cast into a single trait/virtual call.
An alternative I expolored was to do the type erasure internally on `BasePass<ArcComputeCommand<A>>` but I found this quite unwieldy, also this wouldn't cover timestamp writes which are intentionally left unaddressed in this PR.

**Testing**
Added a new test that ensures that resources passed into a ComputePass can be dropped before executing the compute pass.
Proved that the test would crash before these changes and passes with them.

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

### CODE REVIEW

File: wgpu-core/src/lib.rs [hunk 1]
```diff
@@ -301,6 +301,17 @@ macro_rules! gfx_select {
...
+    ($id:expr => $method:ident $params:tt) => {
+        match $id.backend() {
+            wgt::Backend::Vulkan => $crate::gfx_if_vulkan!($method::<$crate::api::Vulkan> $params),
+            wgt::Backend::Metal => $crate::gfx_if_metal!($method::<$crate::api::Metal> $params),
+            wgt::Backend::Dx12 => $crate::gfx_if_dx12!($method::<$crate::api::Dx12> $params),
+            wgt::Backend::Gl => $crate::gfx_if_gles!($method::<$crate::api::Gles> $params),
+            wgt::Backend::Empty => $crate::gfx_if_empty!($method::<$crate::api::Empty> $params),
+            other => panic!("Unexpected backend {:?}", other),
+        }
+    };
```
Comment (Wumpf) [inline wgpu-core/src/lib.rs:314]: remove this again, this was a in-development artifact

File: tests/tests/compute_pass_resource_ownership.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+}
+";
+
+#[gpu_test]
+static COMPUTE_PASS_RESOURCE_OWNERSHIP: GpuTestConfiguration = GpuTestConfiguration::new()
+    .parameters(TestParameters::default().test_features_limits())
+    .run_async(compute_pass_resource_ownership);
+
+async fn compute_pass_resource_ownership(ctx: TestingContext) {
+    let sm = ctx
```
Comment (Wumpf) [inline tests/tests/compute_pass_resource_ownership.rs:91]: on the next pr I have a change that seperates out resource creation. makes this a bit easier to read, let's cherry-pick that in here

File: wgpu/src/backend/wgpu_core.rs [hunk 3]
```diff
@@ -2323,7 +2331,12 @@ impl crate::Context for ContextWgpuCore {
         _bind_group_data: &Self::BindGroupData,
         offsets: &[wgt::DynamicOffset],
     ) {
-        wgpu_compute_pass_set_bind_group(pass_data, index, *bind_group, offsets);
+        if let Err(cause) = pass_data
+            .pass
+            .set_bind_group(&self.0, index, *bind_group, offsets)
+        {
+            self.handle_error_nolabel(&pass_data.error_sink, cause, "ComputePass::set_bind_group");
```
Comment (Wumpf) [inline wgpu/src/backend/wgpu_core.rs:2364]: these all become `handle_error` in the next pr of the train

**Review by cwfitzgerald — APPROVED** [review]
By and large great - I'd be curious is there is any performance difference, but lets ship this

**Comment by Wumpf** [comment]
> By and large great - I'd be curious is there is any performance difference, but lets ship this

Me too! I really want to add a compute pass benchmark and backport it now that you started that work
The next pr in the track has the potential of making things faster again

**Pull request merged**